app/js/lib/foxbox/services/ip-camera.js

```diff
@@ -0,0 +1,86 @@
+'use strict';
+
+import BaseService from './base';
+
+const TYPE = 'IpCamera';
+
+const p = Object.freeze({
+  // Public getter only properties.
+  manufacturer: Symbol('manufacturer'),
+  model: Symbol('model'),
+  name: Symbol('name'),
+
+  // Private getter only properties.
+  adapter: Symbol('adapter'),
+  getters: Symbol('getters'),
+  setters: Symbol('setters'),
+});
+
+export default class IpCameraService extends BaseService {
+  constructor(props, config) {
+    super(props, config);
+
+    // Public getter only properties.
+    this[p.manufacturer] = props.properties.manufacturer;
+    this[p.model] = props.properties.model;
+    this[p.name] = props.properties.name;
+
+    // Private getter only properties.
+    this[p.adapter] = props.adapter;
+    this[p.getters] = props.getters;
+    this[p.setters] = props.setters;
+
+    Object.seal(this);
+  }
+
+  // Public getter only properties.
+  get type() {
+    return TYPE;
+  }
+
+  get manufacturer() {
+    return this[p.manufacturer];
+  }
+
+  get model() {
+    return this[p.model];
+  }
+
+  get name() {
+    return this[p.name];
+  }
+
+  // Public dynamic property getters.
+  getImageList() {
+    return this[p.query]('image_list');
+  }
+
+  getLatestImage() {
+    return this[p.query]('latest image');
+  }
+
+  // Public methods.
+  takeSnapshot() {
```

Comment:
Let me describe what I have in my head so that we can discuss and workout best possible solution :)

I was thinking about the approach when the most of heavy lifting is done in base service, while specific service know just a few things. Let's take a camera service as example. To take a snapshot we have the following method:
```
const operation = this.getOperation(this[p.setters], 'TakeSnapshot');

return this.performSetOperation(operation)
  .then(() => {
    return this.performGetOperation(
      this.getOperation(this[p.getters], 'latest image')
    );
  });
```

Looking at this method I feel unnecessary step - we retrieve operation from base service and then immediately pass it to base service again. That means we can improve it, like this:

```
this.performSetOperation({ kind: 'TakeSnapshot'})
  .then(
    () => this.performGetOperation({ kind: 'latest image' })
  );
```

Basically specific service don't need to have ability to retrieve operation itself, it's too much technical details and too coupled to the current API state. I use ```{ kind: 'TakeSnapshot' }``` cause I was thinking there maybe cases when we want to query channel (I'm trying to use vocabulary from taxonomy where getters and setters are channels) using channel id, in this case we'll have ```{ id: 'clock@adapter' }```.

As in any API we can leverage polymorphous methods to have sensible defaults. Let's say channel kind is the most frequently used channel selector so we can simplify even further:
```
this.performSetOperation('TakeSnapshot')
  .then(() => this.performGetOperation('latest image'));

// Or if we use setter-getter-vocabulary.
// Too confusing though, maybe send/receive or just use current.
this.set('TakeSnapshot').then(() => this.get('latest image'));
``` 
To support this base service should have something like this (semi-pseudo code):
```
set(selector, value) {
  const setter = this[p.getChannel](this[p.service].setters, selector);

  const operationType = this[p.getChannelValueType](setter.kind);

  return this[p.net].fetchJSON(
    `${this[p.net].origin}/api/v${this[p.settings].apiVersion}/channels/set`,
    'PUT',
    [[{ id: setter.id }, { [operationType]: value }]]
  );
}

[p.getChannel](channels, selector) {
  let channelKey;

  if (selector.id) {
    channelKey = selector.id;
  } else if (selector.kind || typeof selector === 'string') {
    // That is optional default I was talking above.
    const channelKind = selector.kind || selector;
    channelKey = Object.keys(channels).find((key) => {
      const channel = channels[key];

      if (typeof channel.kind === 'object') {
        return channel.kind.kind === channelKind;
      }

      return channel.kind === channelKind;
    });
  }

  return channelKey ? channels[channelKey] : null;
}
```

Also theoretically we can pass ```selector``` directly to the service as getter/setter selector - but it's dependency on the API, so I'm really sure.

It's just an idea I had in mind since our last f2f discussion. What do you think? Am I missing something?